### PR TITLE
Set ownership for new folders

### PR DESCRIPTION
**Pull request by zhan9san**
## Proposed Commit Message

```
summary: Set ownership for new folders in Write Files Module

The parent directory would be created automatically if it does not exist. But the ownership of newly-created parent directory would always be root.

With this change, it would be set the same as `owner`.

LP: #1990513
```

## Test Steps

```yaml
write_files:
  - path: /home/ubuntu/.help/file-in.txt
    content: |
      hi in
    owner: 'ubuntu:ubuntu'
    permissions: '0644'
    defer: true
```

With this change, the owner of `/home/ubuntu/.help`  would be `ubuntu`.
Without this change, the owner of `/home/ubuntu/.help`  would be `root`.

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

**Review by holmanb — CHANGES_REQUESTED** [review]
Thanks for the PR and signing the CLA @zhan9san! At a glance this looks pretty good.

I think we should also update the schema definition for `owner` from: `Optional owner:group to chown on the file. Default: root:root.` to something like `Optional owner:group to chown on the file and new directories. Default: root:root.`

**Comment by igalic** [comment]
n.b.: BSDs don't have a `root` group.
easiest portability thing would be to use "0"

**Comment by zhan9san** [comment]
@holmanb Thanks for your suggestion. The owner description in schema has been changed as well. Please help review it and approve ci jobs.

@igalic Thanks for your review. Yes, you are right. BSD is an exception.

But I don't think 0 is a good choice. As the users always specify the name instead of id in owner.

**Comment by igalic** [comment]
ah, in that case, we can just ignore what i said. I thought this was not us providing defaults.
is it's not, we can assume the users know what they're doing ;) and if not, they'll find out on the first run.

**Review by holmanb — CHANGES_REQUESTED** [review]
Thanks, @zhan9san, for the updates. I have one real objection to the current implementation, but I think it should be relatively simple to change using something like `pathlib.Path`'s `parts` or `parent` rather than a recursive walk of all files.

**Comment by holmanb** [comment]
Any updates on this @zhan9san?

**Comment by zhan9san** [comment]
Thanks for your review.
I was a little busy last two weeks.

I'll do it tomorrow

**Comment by zhan9san** [comment]
@holmanb 

Please help review it again.

**Review by holmanb — CHANGES_REQUESTED** [review]

File: cloudinit/util.py [hunk 1]
```diff
@@ -1787,12 +1791,30 @@ def json_dumps(data):
     )
 
 
-def ensure_dir(path, mode=None):
+def get_non_exist_parent_dir(path):
+    """Get non exist parent dir."""
```
[the remark this answers is not in the document]
Comment (zhan9san) [reply cloudinit/util.py]: Thanks for your review.

The return value should be `/usr/a` not `/usr`, which is updated in new commit.

File: cloudinit/util.py [hunk 2]
```diff
@@ -1606,6 +1607,10 @@ def chownbyname(fname, user=None, group=None):
     except KeyError as e:
         raise OSError("Unknown user or group: %s" % (e)) from e
     chownbyid(fname, uid, gid)
+    if recursive and os.path.isdir(fname):
```
[the remark this answers is not in the document]
Comment (zhan9san) [reply cloudinit/util.py]: It's dropped in new commit.

**Review by holmanb — APPROVED** [review]
LGTM! Thanks for this PR. Once CI completes I'll merge.

**Pull request merged**